src/main/java/com/ltonetwork/seasalt/sign/ECDSA.java

```diff
@@ -0,0 +1,140 @@
+        return new KeyPair(publicKey, privateKey);
+    }
+
+    public KeyPair keyPairFromSeed(byte[] seed) {
+        SecureRandom srSeed = new SecureRandom(seed);
+        byte[] privateKey = generatePrivateKey(srSeed);
+        byte[] publicKey = privateToPublic(privateKey);
+        return new KeyPair(publicKey, privateKey);
+    }
+
+    public KeyPair keyPairFromSecretKey(byte[] privateKey) {
+        byte[] publicKey = privateToPublic(privateKey);
+        return new KeyPair(publicKey, privateKey);
+    }
+
+    public byte[] signDetached(byte[] msg, byte[] privateKey, Digest digest) {
+        org.bouncycastle.crypto.Digest bouncyDigest = fetchDigest(digest);
+        ECDSASigner signer = new ECDSASigner(new HMacDSAKCalculator(bouncyDigest));
+        signer.init(true, new ECPrivateKeyParameters(new BigInteger(privateKey), domain));
+        BigInteger[] signature = signer.generateSignature(msg);
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        try {
+            DERSequenceGenerator seq = new DERSequenceGenerator(baos);
+            seq.addObject(new ASN1Integer(signature[0]));
+            seq.addObject(new ASN1Integer(toCanonicalS(signature[1])));
+            seq.close();
+            return baos.toByteArray();
+        } catch (IOException e) {
+            return new byte[0];
+        }
+    }
+
+    public byte[] signDetached(byte[] msg, KeyPair keypair, Digest digest) {
+        return signDetached(msg, keypair.getPrivatekey(), digest);
+    }
+
+    public byte[] signDetached(byte[] msg, byte[] privateKey) {
+        return signDetached(msg, privateKey, Digest.SHA256);
+    }
+
+    public byte[] signDetached(byte[] msg, KeyPair keypair) {
+        return signDetached(msg, keypair.getPrivatekey(), Digest.SHA256);
+    }
+
+    public boolean verify(byte[] msg, byte[] signature, byte[] publicKey) {
+        ASN1InputStream asn1 = new ASN1InputStream(signature);
+        try {
+            ECDSASigner signer = new ECDSASigner();
+            signer.init(false, new ECPublicKeyParameters(curve.getCurve().decodePoint(publicKey), domain));
+
+            DLSequence seq = (DLSequence) asn1.readObject();
+            BigInteger r = ((ASN1Integer) seq.getObjectAt(0)).getPositiveValue();
+            BigInteger s = ((ASN1Integer) seq.getObjectAt(1)).getPositiveValue();
+            return signer.verifySignature(msg, r, s);
+        } catch (Exception e) {
+            return false;
+        } finally {
+            try {
+                asn1.close();
+            } catch (IOException ignored) {
+            }
+        }
+    }
+
+    public boolean verify(byte[] msg, byte[] signature, KeyPair keypair) {
+        return verify(msg, signature, keypair.getPublickey());
+    }
+
+    private byte[] privateToPublic(byte[] privateKey) {
+        return curve.getG().multiply(new BigInteger(privateKey)).getEncoded(true);
+    }
+
+    private org.bouncycastle.crypto.Digest fetchDigest(Digest digest) {
```

Comment:
This doesn't belong here. Add a method on `Digest` to get a bouncy digest object.

Reply:
Removed overall the Enum. Now it requires BouncyCastle Digest to be passed.